### PR TITLE
Bloodbolter

### PR DESCRIPTION
**Pull request by ThatDamnPineapple**
Adds the bloodbolter, a new blood mooncrossbow, aswell as a blood metaball system and some dust to go along with it. Also implements said dust on the steamsaw, and made living bloom animated and drop from blood gazers and blood zombies

### CODE REVIEW

**Review by ScalarVector1 — DISMISSED** [review]
There are a large number of styling problems in the item itself, I also have a number of concerns with the large amount of duplicated code in the lighting buffer class. There is much likely a better implementation of lighting here.

File: Content/Dusts/BloodMetaballDust.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+				dust.scale *= 0.96f;
+
+			if (dust.scale < 0.05f)
+				dust.active = false;
+
+			return false;
+		}
+	}
+
+	public class BloodMetaballDustLight : BloodMetaballDust { public int k = 0; }
```
[the remark this answers is not in the document]
Comment (ThatDamnPineapple) [reply Content/Dusts/BloodMetaballDust.cs:68]: It is called separately on the metaball drawing. It draws with a lighter color

[the remark this answers is not in the document]
Comment (ScalarVector1) [reply Content/Dusts/BloodMetaballDust.cs:69]: What is the point of the ``k`` variable here?

File: Content/Dusts/BloodMetaballDust.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+using System.Collections.Generic;
+using Terraria.Graphics.Effects;
+using Terraria.DataStructures;
+using Terraria.GameContent;
+
+namespace StarlightRiver.Content.Dusts
+{
+	public class BloodMetaballDust : ModDust
+	{
+		public override string Texture => AssetDirectory.Assets + "Invisible";
```
[the remark this answers is not in the document]
Comment (ThatDamnPineapple) [reply Content/Dusts/BloodMetaballDust.cs:24]: fixed

File: Content/Dusts/SmokeDustColor.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            dust.frame = new Rectangle(0, Main.rand.Next(2) * 32, 32, 32);
+            dust.rotation = Main.rand.NextFloat(6.28f);
+        }
+
+        public override Color? GetAlpha(Dust dust, Color lightColor)
+        {
+            Color gray = new Color(25, 25, 25);
+            Color black = Color.Black;
+            Color ret;
+            if (dust.alpha < 120)
```
[the remark this answers is not in the document]
Comment (ThatDamnPineapple) [reply Content/Dusts/SmokeDustColor.cs:25]: Fixed

File: Content/Items/Gravedigger/Materials.LivingBlood.cs [hunk 4]
```diff
@@ -1,9 +1,39 @@
...
 
 namespace StarlightRiver.Content.Items.Gravedigger
 {
     public class LivingBlood : QuickMaterial
     {
-        public LivingBlood() : base("Living Blood", "", 999, 50, 2, AssetDirectory.GravediggerItem) { }
+        public override void Load()
+        {
+			StarlightNPC.ModifyNPCLootEvent += DropLivingBlood;
+        }
```
[the remark this answers is not in the document]
Comment (ThatDamnPineapple) [reply Content/Items/Gravedigger/Materials.LivingBlood.cs:17]: Fixed

File: Content/Items/Gravedigger/Weapons.BloodBolter.cs [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+	public class BloodBolter : ModItem
+	{
+		public override string Texture => AssetDirectory.GravediggerItem + Name;
+
+
+		public override void SetStaticDefaults()
+		{
+			DisplayName.SetDefault("Bloodbolter");
+			Tooltip.SetDefault("Converts wooden arrows into bloodbolts \nBloodbolts impale dead fleshy enemies, exploding them on surfaces");
+
```
[the remark this answers is not in the document]
Comment (ThatDamnPineapple) [reply Content/Items/Gravedigger/Weapons.BloodBolter.cs]: Fixed

File: Content/Items/Gravedigger/Weapons.BloodBolter.cs [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+			Terraria.Audio.SoundEngine.PlaySound(SoundID.Item10, Projectile.Center);
+            for (int i = 0; i < 12; i++)
+            {
+				Vector2 dir = Vector2.Normalize(-Projectile.velocity).RotatedByRandom(0.8f);
+				Dust.NewDustPerfect(Projectile.Center + Main.rand.NextVector2Circular(8, 8), DustID.Blood, dir * Main.rand.NextFloat(7), 0, default, 1.25f);
+            }
+        }
+    }
+
+	public class BloodBolterGNPC : GlobalNPC
```
[the remark this answers is not in the document]
Comment (ThatDamnPineapple) [reply Content/Items/Gravedigger/Weapons.BloodBolter.cs:249]: Fixed

File: Content/Items/Gravedigger/Weapons.BloodBolter.cs [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+			line *= Radius;
+
+			if (Collision.CheckAABBvLineCollision(targetHitbox.TopLeft(), targetHitbox.Size(), Projectile.Center, Projectile.Center + line))
+				return true;
+
+			return false;
+		}
+	}
+
+	public class BloodBolterGoreDestroyer : ModSystem
```
[the remark this answers is not in the document]
Comment (ThatDamnPineapple) [reply Content/Items/Gravedigger/Weapons.BloodBolter.cs]: Done

File: Content/Items/Gravedigger/Weapons.BloodBolter.cs [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+			Projectile.penetrate = -1;
+			Projectile.timeLeft = 10;
+		}
+
+		public override void SetStaticDefaults()
+		{
+			DisplayName.SetDefault("Blood Bolter");
+		}
+
+		public override void AI()
```
[the remark this answers is not in the document]
Comment (ThatDamnPineapple) [reply Content/Items/Gravedigger/Weapons.BloodBolter.cs]: Fixed

File: Helpers/LightingBuffer.cs [hunk 9]
```diff
@@ -311,6 +311,235 @@ public static void DrawWithLighting(Rectangle pos, Texture2D tex, Rectangle sour
             Main.instance.GraphicsDevice.SetVertexBuffer(null);
         }
 
+        public static void DrawRendertargetWithLighting(Rectangle pos, Texture2D tex, Rectangle source, Color color = default)
```
[the remark this answers is not in the document]
Comment (ThatDamnPineapple) [reply Helpers/LightingBuffer.cs]: Fixed

File: Content/Items/Gravedigger/Weapons.BloodBolter.cs [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+			Main.projFrames[Projectile.type] = 7;
+		}
+
+		public override void SetDefaults()
+		{
+			Projectile.friendly = false;
+			Projectile.tileCollide = false;
+			Projectile.Size = new Vector2(32, 32);
+			Projectile.penetrate = -1;
+		}
```
[the remark this answers is not in the document]
Comment (ThatDamnPineapple) [reply Content/Items/Gravedigger/Weapons.BloodBolter.cs:95]: where?

File: Content/Tiles/Vitric/VitricBossAltar.cs [hunk 11]
```diff
@@ -340,7 +340,7 @@ private bool checkIfDrawReflection()
             int off = (int)(BarrierProgress / 120f * tex.Height);
             int off2 = (int)(BarrierProgress / 120f * texTop.Width / 2);
 
-            LightingBufferRenderer.DrawWithLighting(new Rectangle((int)center.X - 790 - (int)Main.screenPosition.X, (int)center.Y - off - 16 - (int)Main.screenPosition.Y, tex.Width, off), tex, new Rectangle(0, 0, tex.Width, off), default);
+            LightingBufferRenderer.DrawWithLighting(new Rectangle(((int)center.X - 790 - (int)Main.screenPosition.X) - (tex.Width / 2), ((int)center.Y - off - 16 - (int)Main.screenPosition.Y) - (tex.Height / 2), tex.Width, off), tex, new Rectangle(0, 0, tex.Width, off), default);
```
Comment (ScalarVector1) [inline Content/Tiles/Vitric/VitricBossAltar.cs]: Why does this behavior need to change? This is concerning -- are the lighting buffer changes here breaking?

**Review by ScalarVector1 — APPROVED** [review]
Going to approve for now, still a bit hesitant to merge this with lighting buffer changes to make sure they're 100% non-breaking.

**Pull request merged**